ring/consul_client.go

```diff
@@ -206,6 +212,65 @@ func (c *consulClient) WatchPrefix(prefix string, factory func() interface{}, do
 	}
 }
 
+// WatchKey will watch a given key in consul for changes. When the value
+// under said key changes, the f callback is called with the deserialised
+// value. To construct the deserialised value, a factory function should be
+// supplied which generates an empty struct for WatchPrefix to deserialise
+// into. Values in Consul are assumed to be JSON. This function blocks until
+// the done channel is closed.
+func (c *consulClient) WatchKey(key string, factory InstanceFactory, done chan struct{}, f func(interface{}) bool) {
+	const (
+		initialBackoff = 1 * time.Second
+		maxBackoff     = 1 * time.Minute
+	)
+	var (
+		backoff = initialBackoff / 2
+		index   = uint64(0)
+	)
+	for {
+		select {
+		case <-done:
+			return
+		default:
+		}
+		kvp, meta, err := c.kv.Get(key, &consul.QueryOptions{
+			RequireConsistent: true,
+			WaitIndex:         index,
+			WaitTime:          longPollDuration,
+		})
+		if err != nil {
+			log.Errorf("Error getting key %s: %v", key, err)
+			backoff = backoff * 2
+			if backoff > maxBackoff {
+				backoff = maxBackoff
+			}
+			select {
+			case <-done:
+				return
+			case <-time.After(backoff):
+				continue
+			}
+		}
+		backoff = initialBackoff
+		if index == meta.LastIndex {
+			continue
+		}
+		index = meta.LastIndex
```

Comment:
It took me a while to grok that `LastIndex` here means "the index we got last time" rather than "the last possible index", which made the overall logic a bit confusing. 

Perhaps add a comment saying something like "Skip if the index is the same as last time, because there's no new data"? WDYT?